### ISSUE TITLE
Question: How to create a Var from MmapedSafetensors

### ISSUE THREAD

**Issue by EricLBuehler** [huggingface/candle#1185]
Hello everybody,

I was wondering how to create a Var instance from an `MMapedSafetensors` `TensorView`. I have tried using `candle_core::Var::from_slice(tensor.data(), tensor.shape(), &device)?`, but I get the error:

`Error: Shape mismatch, got buffer of size 90177536 which is compatible with shape [11008, 4096]`.

Is there a better way to do this? 

In addition, I notice the buffer is of type `u8`, which is definitely not the data type the safetensors should be decoded as. Where can I find how `VarBuilder` does this?

**In summary, I have 2 questions:**
- How to decode a `TensorView` into a `Var`?
- Or, if the above is not feasible, how does `VarBuilder` do this?

**Comment by EricLBuehler** [comment]
Update: This code:

```rust
let map = VarMap::new();
            {
                let mut ws = map.data().lock().unwrap();

                let tensors =
                    unsafe { candle_core::safetensors::MmapedSafetensors::multi(&filenames)? };
                for (name, tensor) in tensors.tensors() {
                    let tensor = BufferedSafetensors::new(tensor.data().to_vec())
                        .unwrap()
                        .load(&name, &device)
                        .unwrap();
                    ws.insert(name, candle_core::Var::from_tensor(&tensor)?);
                }
            }
            let vb = candle_nn::VarBuilder::from_varmap(&map, candle_core::DType::F32, &device);
```

Panics with the following error: `thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: SafeTensor(HeaderTooLarge)'`.

@LaurentMazare, is there a way to fix this by using what `VarBuilder` does? Please note that I need to load the safetensors into a `VarMap` for the insertion of LoRA layers.

**Comment by LaurentMazare** [comment]
- `SafeTensor(HeaderTooLarge)` is likely that the file that your loading is not in safetensors format.
- You can create a `VarMap`, initializes the variables in the model creation code, and then load the values from a safetensor files. That's what is done in the training examples in this repo and maybe it would be the best for you.
- Otherwise you may want to use `Var::from_tensor`, not sure why you started with `Var::from_slice` but this one is a fairly low-level "expert" api so probably better to avoid it if you can.

**Comment by EricLBuehler** [comment]
Thank you! This code resolved my issue:
```rust
let map = VarMap::new();
{
    let mut ws = map.data().lock().unwrap();

    let tensors =
        unsafe { candle_core::safetensors::MmapedSafetensors::multi(&filenames)? };
    for (name, _) in tensors.tensors() {
        let tensor = tensors.load(&name, &device)?;
        ws.insert(name, Var::from_tensor(&tensor)?);
    }
}
let vb = candle_nn::VarBuilder::from_varmap(&map, candle_core::DType::F32, &device);
```
